### ISSUE TITLE
DataStore doesn't emit `syncQueriesReady` or `ready` events after first-time sync

### ISSUE THREAD

**Issue by albemala** [aws-amplify/amplify-ios#1302]
**Describe the bug**

I'm using Amplify with Auth, Storage, API and DataStore in my app. The first time I run the app after installing it on a device, Amplify does a full DataStore sync with the back-end. When DataStore finishes synchronizing, it should emit a `syncQueriesReady` or `ready` event, but it doesn't. If I close the app and reopen it, DataStore synchronizes correctly and the 2 events are emitted correctly.

**To Reproduce**

1. Setup Auth, Storage, API and DataStore (both app and back-end)
2. Install the app
3. The first time you install the app, there is no user logged in. DataStore doesn't synchronize at this point.
4. Log in with a user
5. After a successful login, DataStore starts to sync with the back-end.

**Expected behavior**

When DataStore finishes synchronizing, it should emit a `syncQueriesReady` or `ready` event, but it doesn't.

**Environment(please complete the following information):**

 - Amplify Framework Version: 1.9.3
 - Amplify Categories:
   - [ ] Analytics
   - [x] API
   - [x] Auth
   - [x] DataStore
   - [ ] Predictions
   - [x] Storage
 - Dependency Manager: Cocoapods
 - Swift Version: 5.x
 - CLI Version: 4.50.2
 - Xcode Version: 12.5.1

**Is this a regression (i.e. was this working before a version upgrade?)**

I'm not sure. I _think_ it's a recent issue, but I also suspect it's related to the back-end database that is growing and it takes more time to synchronize.

**Device Information (please complete the following information):**

 - Device: iPhone 12 Pro
 - iOS Version: iOS 14.6
 - Specific to simulators: No

**Additional context**

I hope my explanation is clear enough. If not, please let me know and I'll provide more information. Thank you for your help!

**Comment by diegocstn** [comment]
hi @albemala,
thanks for reaching out.
How does your schema look like? Do all your models have associated one or more @auth rules?
I'd be also helpful to know how you're triggering DataStore initial sync after the user has logged in (i.e. by calling `Amplify.DataStore.start` or by just querying one of the models).

**Comment by albemala** [comment]
Thank you @diegocstn for the prompt reply!

> How does your schema look like? Do all your models have associated one or more @auth rules?

All types have at least one `@auth` rule associated, with 2 exceptions: `S3Object` and a `Query` type we defined to trigger lambda functions using the Amplify API.

> how you're triggering DataStore initial sync after the user has logged in

By querying one of the models.

**Comment by lawmicha** [comment]
Do you have sample of your schema we can use to reproduce this? Are you using `@connection` anywhere? Do you see any relevant logs indicating an error with the sync process? You can turn on verbose logging with ` Amplify.Logging.logLevel = .verbose`

**Comment by albemala** [comment]
@lawmicha Sorry I cannot share the schema.

Yes we are using `@connection`s in a few places.

I tried to turn on verbose logging, but I don't see any relevant error from Amplify. `modelSynced` events are correctly emitted. But `syncQueriesReady` or `ready` events are never emitted the first time I run the app.

**Comment by diegocstn** [comment]
@albemala does the number of `modelSynced` events match the number of models in your schema?
We've recently merged a [fix](https://github.com/aws-amplify/amplify-ios/pull/1299) to specifically address an issue with the Hub events when one or more models have @auth rules and there's no user signed in.

**Comment by anatomybook** [comment]
I'm seeing a similar issue. Sometimes both events don't get emitted. I've yet to find out why/when exactly this happens. Also with verbose logging I see nothing out of the ordinary.
For me it seems to occur more frequently when there are not many live items in the tables, but many deleted ones. Since my app modifies the items on a daily basis, I only have a day to investigate each time. Once it does occur, however, I can reproduce it. So far I haven't found the time to really dig into this.

**Comment by diegocstn** [comment]
@anatomybook can you share more details about your schema?
In particular, we'd be interested to know if any of your models have @auth rules associated or @connection to other models.

**Comment by anatomybook** [comment]
Hello @diegocstn ,

I've included a redacted version below. The information you were looking for should still be present. I merely removed most data fields (Ints, Strings, Enums) of the types and renamed the rest. I have not removed any type or any connection.

```
type TypeA @model @auth(rules: [{ allow: owner }])
{
  id: ID!

  typeBs: [TypeB] @connection(name: "LinkF")
  groups: [TypeF] @connection(name: "LinkA")
  typeDs: [TypeD] @connection(name: "LinkE")
}

type TypeB @model @auth(rules: [{ allow: owner }])
{
  id: ID!

  typeA: TypeA @connection(name: "LinkF")
  typeEs: [TypeE] @connection(name: "LinkG")
  filenames: [TypeC] @connection(name: "LinkH")
}

type TypeC @model @auth(rules: [{ allow: owner }])
{
  id: ID!
  sThreeKey: String!
  lastLookedAt: AWSDateTime
  remarks: String
  details: String
  receipt: TypeB @connection(name: "LinkH")
}

type TypeE @model @auth(rules: [{ allow: owner }])
{
  id: ID!

  parent: TypeB @connection(name: "LinkG")
  groups: [TypeK] @connection(keyName: "typeListNext", fields: ["id"])
}

type TypeK @model(queries: null) @auth(rules: [{ allow: owner }])
@key(name: "typeListNext", fields: ["typeNextId", "typeTwoId"])
@key(name: "typeFListA", fields: ["typeTwoId", "typeNextId"])
{
  id: ID!
  typeNextId: ID!
  typeTwoId: ID!
  typeE: TypeE! @connection(fields: ["typeNextId"])
  typeF: TypeF! @connection(fields: ["typeTwoId"])
}

type TypeF @model @auth(rules: [{ allow: owner }])
{
  id: ID!

  typeA: TypeA @connection(name: "LinkA")
  typeGs: [TypeG] @connection(name: "LinkB")
  typeHs: [TypeH] @connection(name: "LinkC")
  statusChangeTime: [TypeL] @connection(name: "LinkD")
  typeDs: [TypeI] @connection(keyName: "typeFList", fields: ["id"])
  typeEs: [TypeK] @connection(keyName: "typeFListA", fields: ["id"])
}

type TypeG @model @auth(rules: [{ allow: owner }])
{
  id: ID!

  typeF: TypeF @connection(name: "LinkB")
}

type TypeL @model @auth(rules: [{ allow: owner }])
{
  id: ID!
  statusChangedTo: TypeFShoppingStatusType!
  statusChangeTime: AWSDateTime!
  typeF: TypeF @connection(name: "LinkD")
}

type TypeH @model @auth(rules: [{ allow: owner }])
{
  id: ID!

  typeF: TypeF @connection(name: "LinkC")
}

type TypeI @model(queries: null) @auth(rules: [{ allow: owner }])
@key(name: "typeDList", fields: ["typeOneId", "typeTwoId"])
@key(name: "typeFList", fields: ["typeTwoId", "typeOneId"])
{
  id: ID!
  amount: Float
  currentStatus: TypeFShoppingStatusType
  typeOneId: ID!
  typeTwoId: ID!
  typeD: TypeD! @connection(fields: ["typeOneId"])
  typeF: TypeF! @connection(fields: ["typeTwoId"])
}

type TypeD @model @auth(rules: [{ allow: owner }])
{
  id: ID!
  name: String!

  typeA: TypeA @connection(name: "LinkE")
  groups: [TypeI] @connection(keyName: "typeDList", fields: ["id"])
}

```

Please let me know if you have any further questions.

**Comment by albemala** [comment]
@diegocstn good news! I've tested v1.12.0 and the issue seems to be fixed 🎉 

We'll run some more tests this week and I'll confirm as soon as possible.

**Comment by anatomybook** [comment]
I upgraded to c622147b17b198cdc9d15a4ff049d7aa520139e1 and no change for me. Sometimes certain Models don't get synced without any error message. In the scheme, I provided above, it's currently `TypeF` that doesn't sync.

**Comment by anatomybook** [comment]
Hello @diegocstn ,

I just spent some time digging into this, since today the issue occurred again.

To me it looks like there's a problem in the order in which `ModelSyncedEventEmitter.onReceiveSyncOperationEvent` and `ModelSyncedEventEmitter.onReceiveReconciliationEvent` are called.
As far as I can tell, the idea is that first `onReceiveSyncOperationEvent` is called with `.finished` setting `initialSyncOperationFinished` to true. Then the synced event is dispatched from `onReceiveReconciliationEvent` if `initialSyncOperationFinished && reconciledReceived == recordsReceived`.  For most of my models, that's the case. However, I found with the problematic model the order is broken, i.e. first `onReceiveReconciliationEvent` is called which does nothing because `initialSyncOperationFinished` is still false. Then the call to `onReceiveSyncOperationEvent` arrives setting `initialSyncOperationFinished` to true. However, this is useless since `onReceiveReconciliationEvent` is never called again.

Does this make sense?

I didn't dig any deeper regarding _why_ this happens. But perhaps this already helps?! If you have any questions, please let me know.

**Comment by anatomybook** [comment]
Add-on:

I added the snippet:
```
            if reconciledReceived == recordsReceived {
                dispatchModelSyncedEvent()
            }
```
beneath
```
if recordsReceived == 0 {
                dispatchModelSyncedEvent()
            }
```
in `ModelSyncedEventEmitter.onReceiveSyncOperationEvent` and it fixed the issue for me. However, I've no idea how many other things broke in the process.

**Comment by diegocstn** [comment]
@albemala glad to hear that, let us know how the test goes!

@anatomybook thanks for the details, we're gonna dive deep on that.

**Comment by albemala** [comment]
@diegocstn We did some more tests last week and everything seems to be working fine now, after moving to v1.12.0. As for me, you can close this ticket. Thank you for your help!

**Comment by anatomybook** [comment]
Could this be related to this issue?

https://github.com/aws-amplify/amplify-ios/issues/1331

**Comment by lawmicha** [comment]
The original issue seems to have been fixed in v1.12.0. @anatomybook could you try the latest version https://github.com/aws-amplify/amplify-ios/releases/tag/v1.15.5 which contains some changes to how the events are sent back to the app. Please open another another issue specific to your situation if something still isn't working for you
